### DISCUSSION TITLE
How to pass secrets in Supabase migration scripts?

### DISCUSSION THREAD

**Discussion by abdullahsych** [supabase/supabase#19229] [Questions]
I am working on a migration script that creates custom roles with login and password. Since migration scripts are version controlled, and are also stored in the `supabase_migrations` schema in plain text, typing passwords directly in the migration scripts is obviously a big NO. 

Why am I creating roles in the migration script? I think it comes very handy to test everything together. Other migration tools like [Flyway](https://documentation.red-gate.com/fd/placeholders-224919737.html) support placeholders that can reference secrets which are replaced at runtime. Does Supabase migrations provide similar functionality?

As an alternative I was thinking of creating the roles with `NULL` passwords in the migration script so that all privileges could be version controlled and tested. Passwords would then be created manually after migrations are pushed to production. Any thoughts?

**Comment by jfrosch** [comment]
For this exact reason , I've decided not to use Supabase migrations and instead have a dedicated Flyway project. Maintaining a separate project sucks when there's an otherwise decent database migration capability built in. I'm hopeful the new Supabase Vault service will offer an integration path for the migration scripts; e.g. something like:

```
create user rw_user with password '${appVault.db_rw_password}';
create user ro_user with password '${appVault.db_ro_password}';
```

**Reply by abdullahsych** [reply]
Yeah that'd be very handy. I am hoping someone from Supabase can shed light on this and suggest a work around for the time being.

For this very reason, we are considering making a switch to Flyway. Any issues, gotchas, making Flyway work with Supabase?

**Reply by jfrosch** [reply]
There are a few.

If working with custom schemas, you'll need to jump through some hoops. The move to eliminate `superuser` role from Supabase customers introduces a bit of pain. So you need to grant the `postgres`user to access the schema. If using the cloud, you'll got to DB settings and add the schema(s). If self-hosted, you'll need to add the schemas to the config.toml.

This was the biggest pain point and makes Supabase a bit unfriendly to SaaS implementation where each client gets their own schema. (I didn't investigate the Management API to see if it would yield a programmatic way to handle this.) Once you have the config set up, a Flyway migration will need to include these statements **_for each custom schema_**:

```
GRANT USAGE ON SCHEMA ${db_schema} TO anon, authenticated, service_role;
GRANT ALL ON ALL TABLES IN SCHEMA ${db_schema} TO anon, authenticated, service_role;
GRANT ALL ON ALL ROUTINES IN SCHEMA ${db_schema} TO anon, authenticated, service_role;
GRANT ALL ON ALL SEQUENCES IN SCHEMA ${db_schema} TO anon, authenticated, service_role;
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA ${db_schema} GRANT ALL ON TABLES TO anon, authenticated, service_role;
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA ${db_schema} GRANT ALL ON ROUTINES TO anon, authenticated, service_role;
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA ${db_schema} GRANT ALL ON SEQUENCES TO anon, authenticated, service_role;
```

Otherwise, just getting the Flyway placeholders defined from ENV VARs was the only heavier lift.

Here's my [build.gradle Gist](https://gist.github.com/jfrosch/077f74c9bb5e8970af9ed14fd84fe2cd). An extra bonus is some code to read a `.env.<environment>` file and store the vars as `project` vars that can then be used in placeholders. I'm using a Gradle project with the [Flyway plugin](https://documentation.red-gate.com/fd/gradle-task-184127407.html), so you would need to adapt if you're using Maven or something else.

I'm in early days of using Supabase, but I have successfully executed `flywayMigrate` and `flywayClean` against a local and cloud installation of Supabase.

**Reply by abdullahsych** [reply]
Thanks for sharing this! 

We are not using the PostgREST API exposed by Supabase and our application connects to Postgres directly so we'll keep `anon`, `authenticated` and any other supabase roles away from our schemas. We'll probably use the `postgres` user to run the flyway migrations like Supabase does. Best of luck on your Supabase project!

**Comment by burggraf** [comment]
From our team:
For role passwords, I would recommend using encrypted values in migration file instead of plain text https://stackoverflow.com/questions/17429040/creating-user-with-encrypted-password-in-postgresql

But if they you separate passwords for staging / production projects, referencing secret from env var could be better.
